pandas/io/common.py

```diff
@@ -276,12 +276,17 @@ def _get_filepath_or_buffer(
         fsspec_mode += "b"
 
     if isinstance(filepath_or_buffer, str) and is_url(filepath_or_buffer):
-        # TODO: fsspec can also handle HTTP via requests, but leaving this unchanged
-        if storage_options:
-            raise ValueError(
-                "storage_options passed with file object or non-fsspec file path"
-            )
-        req = urlopen(filepath_or_buffer)
+        # TODO: fsspec can also handle HTTP via requests, but leaving this
+        # unchanged. using fsspec appears to break the ability to infer if the
+        # server responded with gzipped data
+        storage_options = storage_options or dict()
+        # waiting until now for importing to match intended lazy logic of
+        # urlopen function defined elsewhere in this module
+        import urllib.request
+
+        # assuming storage_options is to be interpretted as headers
+        req_info = urllib.request.Request(filepath_or_buffer, headers=storage_options)
```

Comment:
are there more (important) keyword arguments that `Request` takes? If yes, it might be more future-proof to do:
`urllib.request.Request(filepath_or_buffer, **storage_options)`

Reply:
I like that thought, though from reviewing the options I don't think there's much to be gained by exposing the other keyword arguments.  It looks like the other kwargs are there mainly to post data to a url, but that doesn't look to be possible inside the ``_get_filepath_or_buffer`` function due to what is available when that function is called.  Here's the documentation link in case anyone has a different take: [https://docs.python.org/3/library/urllib.request.html#urllib.request.Request](url)

Another less restrictive option might be to change it so that ``storage_options = {"headers": {"User-Agent": "pandas"}}`` leading to ``urllib.request.Request(filepath_or_buffer, headers=storage_options["headers"])``.  This means the options could be used for other functions in ``_get_filepath_or_buffer``.

Both may be a little cumbersome on the user, since I see little benefit in this function in the http(s) case, but it may provide some flexibility that would be nice elsewhere.  Easy enough to do either option and change the docs, or leave it as is.  Got any particular preference among those options?  Thanks!

Reply:
`is_url` strings that are not `is_fsspec_url`, did not use any storage_options before this PR. All options could go into the header/to urllib.request.Request.
If we do something like `Request(filepath_or_buffer, headers=storage_options["headers"])` then we could have other options that are not used but we don't raise about them (or urllib doesn't complain about the provided arguments).

Reply:
As far as I know, there are common uses for some header keys but there's also not really a limit on what header keys can be used (ie. I think all keys are accepted in headers).  To back this up the ``urllib.request.Request`` documentation says that it will call the ``urllib.request.Request.add_header`` method on each key value pair, and from looking at that method and the ``__init__`` method, I don't see any restriction on keys allowed in the header dictionary either.

I think once I am free later tonight I will go ahead and change it so that the usage and documentation lines up with ``Request(filepath_or_buffer, headers=storage_options["headers"])``